Support channel level application health check options

zhoukangsheng opened the pull request:
### What problem does this PR solve?
**Current Application-Layer Health Check Mechanism**
The current application-layer health check mechanism is configured via the -health_check_path flag. If this flag is left empty, the health check is disabled. However, since gflag is a global variable, once set, it enables the application-layer health check for all downstream services accessed by the service.

For complex services that may interact with dozens or even hundreds of downstream services (some of which might not be brpc-based and could instead be Go/Java services), not all downstream services necessarily support health check requests. Even for those that do, the health check request paths (URLs) may vary across services. Therefore, it is necessary to support channel-level configuration for application-layer health check parameters.

Issue Number:

Problem Summary:

### What is changed and the side effects?
no

Changed:
**About This PR**
This PR maintains backward compatibility with the existing gflag-based configuration for health check paths and timeouts while introducing support for setting health check parameters via members of ChannelOptions (hc_option). Parameters in ChannelOptions (hc_option) take precedence over the global gflag settings. If a parameter is not specified in ChannelOptions, the corresponding gflag value will be used as the fallback.

Side effects:
- Performance effects:  no
- Breaking backward compatibility:  no
---
### Check List:
- Please make sure your changes are compilable.
- When providing us with a new feature, it is best to add related tests.
- Please follow [Contributor Covenant Code of Conduct](https://github.com/apache/brpc/blob/master/CODE_OF_CONDUCT.md).

wwbmmm commented on src/brpc/details/naming_service_thread.h:
```diff
@@ -45,11 +45,14 @@ struct GetNamingServiceThreadOptions {
     bool succeed_without_server;
     bool log_succeed_without_server;
     bool use_rdma;
+    int32_t health_check_timeout_ms;
```
Can change these two members to a single HealthCheckOption member

wwbmmm commented on src/brpc/channel.h:
```diff
@@ -143,6 +143,18 @@ struct ChannelOptions {
+    // Its priority is higher than FLAGS_health_check_path
+    // When it is not set, FLAGS_health_check_path will take effect
+    // Default: ""
+    std::string health_check_path;
```
Can change these two members to a single HealthCheckOption member

wwbmmm commented on src/brpc/socket.h:
```diff
@@ -266,6 +266,8 @@ struct SocketOptions {
     // one thread at any time.
     void (*on_edge_triggered_events)(Socket*){NULL};
     int health_check_interval_s{-1};
+    std::string health_check_path;
```
Can change these two members to a single HealthCheckOption member

wwbmmm commented on src/brpc/socket_map.h:
```diff
@@ -44,6 +44,11 @@ inline bool operator!=(const ChannelSignature& s1, const ChannelSignature& s2) {
 
+struct HealthCheckOption {
+    std::string health_check_path;
+    int32_t health_check_timeout_ms = 500;
```
better initialize like this: `int32_t health_check_timeout_ms{500};`

wwbmmm commented on src/brpc/channel.cpp:
```diff
@@ -388,6 +408,10 @@ int Channel::Init(const char* ns_url,
     ns_opt.log_succeed_without_server = _options.log_succeed_without_server;
     ns_opt.use_rdma = _options.use_rdma;
     ns_opt.channel_signature = ComputeChannelSignature(_options);
```
maybe these hc options should be included in ComputeChannelSignature

wwbmmm commented:
Thank you for contributing! This is a useful feature.

zhoukangsheng commented:
> Thank you for contributing! This is a useful feature.
> I have modified the title to English. Since this is an Apache project and we'd better using Engish in PR discussions.

Thank you for your code review very much. I have modified it according to your suggestion. Please check if there is anything else that needs to be modified.

wwbmmm commented on src/brpc/channel.h:
```diff
@@ -36,6 +36,7 @@
 #include "brpc/retry_policy.h"
 #include "brpc/backup_request_policy.h"
 #include "brpc/naming_service_filter.h"
+#include "brpc/socket.h"
```
I think the `HealthCheckOption`  is better to be defined in a separate header file, like `ChannelSSLOptions` defined in `ssl_options.h`. So that you can avoid importing the unrelated classes and headers in `socket.h` here.

zhoukangsheng replied:
done

wwbmmm commented on src/brpc/socket.h:
```diff
@@ -980,6 +992,8 @@ friend void DereferenceSocket(Socket*);
     int _tcp_user_timeout_ms;
 
     HttpMethod _http_request_method;
+    std::string _health_check_path;
```
These can be replaced into `HealthCheckOption` too.

wwbmmm commented on src/brpc/details/health_check.cpp:
```diff
@@ -65,6 +57,8 @@ class OnAppHealthCheckDone : public google::protobuf::Closure {
     SocketId id;
     int64_t interval_s;
     int64_t last_check_time_ms;
+    int32_t health_check_timeout_ms;
```
These can be replaced into HealthCheckOption too.

zhoukangsheng commented:
![ut](https://github.com/user-attachments/assets/93820947-8921-4683-ac11-f1757ff6b145)
The implementation of this UT appears to fail randomly. Increasing the number of executions of SelectServer can, to some extent, make the probability of IP selection closer to the proportion of the set weight to the total weight.

wwbmmm commented:
LGTM
